File: Entities/SteamEmitter.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.PuzzleIslandHelper.Entities
{
    [CustomEntity("PuzzleIslandHelper/SteamEmitter")]
    [Tracked]
    public class SteamEmitter : Entity
    {
        public string Flag;
        private Image image;
        public enum Directions
        {
            Right, Up, Left, Down
        }
        public float Angle;
        private float delay;
        private Vector2 emitOffset;
        private float timer;
        private float interval;
        public SteamEmitter(EntityData data, Vector2 offset) : base(data.Position + offset)
        {
            
            Angle = (MathHelper.Pi / -2f) * (float)data.Enum<Directions>("direction");
            image = new Image(GFX.Game["objects/PuzzleIslandHelper/cap"]);
            image.Rotation = Angle;
            image.CenterOrigin();
            image.Position += image.HalfSize();
            Add(image);
            interval = data.Float("interval", 0.3f);
            Flag = data.Attr("flag");
            Collider = new Hitbox(8, 8);
            emitOffset = Calc.AngleToVector(Angle, 6);
            image.RenderPosition = Center + Calc.AngleToVector(Angle, 8);

        }
        public override void Added(Scene scene)
        {
            base.Added(scene);
            delay = Calc.Random.NextFloat();
            timer = delay + interval;
        }
        public override void Update()
        {
            base.Update();
            if (Scene is not Level level || !Scene.OnInterval(interval) || !Flag.GetFlag()) return;
            level.ParticlesFG.Emit(ParticleTypes.Steam, Center + emitOffset, Color.White * Calc.Random.Range(0.5f, 1.1f), Angle);
        }
    }
}

File: Loenn/entities/steamEmitter.lua
local steamEmitter = {}
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

steamEmitter.justification = { 0, 0 }

local directions = {"Up","Down","Left","Right"}
steamEmitter.name = "PuzzleIslandHelper/SteamEmitter"

steamEmitter.depth = -100000
steamEmitter.texture = "objects/PuzzleIslandHelper/steamEmitter/lonn"
steamEmitter.placements =
{
    {
        name = "Steam Emitter",
        data = {
             direction = "Up",
             interval = 0.1,
             flag = ""
        }
    }
}
steamEmitter.fieldInformation =
{
    direction = 
    {
        options = directions,
        editable = false
    }
}
return steamEmitter